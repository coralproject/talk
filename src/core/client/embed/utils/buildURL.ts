export default function buildURL({
  protocol = window.location.protocol,
  hostname = window.location.hostname,
  port = window.location.port,
  pathname = window.location.pathname,
  search = window.location.search,
  hash = window.location.hash,
} = {}) {
  if (search && search[0] !== "?") {
    search = `?${search}`;
  } else if (search === "?") {
    search = "";
  }
  return `${protocol}//${hostname}${
    port ? `:${port}` : ""
  }${pathname}${search}${hash}`;
}
