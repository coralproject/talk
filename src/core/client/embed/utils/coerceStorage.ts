import { createInMemoryStorage } from "coral-framework/lib/storage";

function coerceStorage(type: "localStorage" | "sessionStorage"): Storage {
  let storage: Storage;
  try {
    storage = window[type];
  } catch {
    // eslint-disable-next-line no-console
    console.debug(`${type} not available, use in-memory replacement`);

    storage = createInMemoryStorage();
  }

  return storage;
}

export default coerceStorage;
