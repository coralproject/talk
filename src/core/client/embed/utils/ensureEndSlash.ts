export default function ensureEndSlash(p: string) {
  return p.match(/\/$/) ? p : `${p}/`;
}
