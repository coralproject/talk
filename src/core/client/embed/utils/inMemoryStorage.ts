class InMemoryStorage implements Storage {
  private readonly data = new Map<string, string>();

  public get length() {
    return this.data.size;
  }

  public clear() {
    this.data.clear();
  }

  public key(n: number) {
    const keys = [...this.data.keys()];
    if (keys.length < n || n < 0) {
      return null;
    }

    return keys[n];
  }

  public getItem(key: string): string | null {
    return this.data.get(key) || null;
  }

  public removeItem(key: string): void {
    this.data.delete(key);
  }

  public setItem(key: string, value: string): void {
    this.data.set(key, value);
  }
}

export default InMemoryStorage;
