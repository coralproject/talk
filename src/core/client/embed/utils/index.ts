export { default as buildURL } from "./buildURL";
export { default as ensureEndSlash } from "./ensureEndSlash";
export { default as ensureNoEndSlash } from "./ensureNoEndSlash";
export { default as startsWith } from "./startsWith";
export { default as prefixStorage } from "./prefixStorage";
export { default as parseHashQuery } from "./parseHashQuery";
