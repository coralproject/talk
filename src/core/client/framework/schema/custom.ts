/**
 * TODO: (cvle) This file is a workaround to have Relay compatible enum types.
 * This should be generated by `graphql-schema-typescript`.
 */

import { RelayEnumLiteral } from "../types";
import {
  GQLCOMMENT_FLAG_DETECTED_REASON,
  GQLCOMMENT_FLAG_REASON,
  GQLCOMMENT_FLAG_REPORTED_REASON,
  GQLCOMMENT_SORT,
  GQLCOMMENT_STATUS,
  GQLLOCALES,
  GQLMODERATION_MODE,
  GQLSTORY_STATUS,
  GQLUSER_AUTH_CONDITIONS,
  GQLUSER_ROLE,
  GQLUSER_STATUS,
} from "./__generated__/types";

export type GQLUSER_ROLE_RL = RelayEnumLiteral<typeof GQLUSER_ROLE>;
export type GQLUSER_STATUS_RL = RelayEnumLiteral<typeof GQLUSER_STATUS>;
export type GQLCOMMENT_FLAG_DETECTED_REASON_RL = RelayEnumLiteral<
  typeof GQLCOMMENT_FLAG_DETECTED_REASON
>;
export type GQLCOMMENT_FLAG_REASON_RL = RelayEnumLiteral<
  typeof GQLCOMMENT_FLAG_REASON
>;
export type GQLCOMMENT_FLAG_REPORTED_REASON_RL = RelayEnumLiteral<
  typeof GQLCOMMENT_FLAG_REPORTED_REASON
>;
export type GQLCOMMENT_SORT_RL = RelayEnumLiteral<typeof GQLCOMMENT_SORT>;
export type GQLCOMMENT_STATUS_RL = RelayEnumLiteral<typeof GQLCOMMENT_STATUS>;
export type GQLLOCALES_RL = RelayEnumLiteral<typeof GQLLOCALES>;
export type GQLMODERATION_MODE_RL = RelayEnumLiteral<typeof GQLMODERATION_MODE>;
export type GQLSTORY_STATUS_RL = RelayEnumLiteral<typeof GQLSTORY_STATUS>;
export type GQLUSER_AUTH_CONDITIONS_RL = RelayEnumLiteral<
  typeof GQLUSER_AUTH_CONDITIONS
>;
