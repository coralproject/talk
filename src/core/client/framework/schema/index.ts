export * from "./__generated__/types";
export * from "./custom";
