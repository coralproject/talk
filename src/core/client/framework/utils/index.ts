export { default as buildURL } from "./buildURL";
export { default as parseURL } from "./parseURL";
export { default as modifyQuery } from "./modifyQuery";
