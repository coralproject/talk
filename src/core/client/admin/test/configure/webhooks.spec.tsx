import { noop } from "lodash";

import { pureMerge } from "coral-common/utils";
import { GQLResolver } from "coral-framework/schema";
import {
  act,
  createResolversStub,
  CreateTestRendererParams,
  replaceHistoryLocation,
  wait,
  waitForElement,
  within,
} from "coral-framework/testHelpers";

import create from "../create";
import { settings, users } from "../fixtures";

beforeEach(async () => {
  replaceHistoryLocation("http://localhost/admin/configure/webhooks");
});

const viewer = users.admins[0];

async function createTestRenderer(
  params: CreateTestRendererParams<GQLResolver> = {}
) {
  const { testRenderer, context } = create({
    ...params,
    resolvers: pureMerge(
      createResolversStub<GQLResolver>({
        Query: {
          settings: () => settings,
          viewer: () => viewer,
        },
      }),
      params.resolvers
    ),
    initLocalState: (localRecord, source, environment) => {
      if (params.initLocalState) {
        params.initLocalState(localRecord, source, environment);
      }
    },
  });

  return await act(async () => {
    const container = await waitForElement(() =>
      within(testRenderer.root).getByTestID("webhooks-container")
    );

    return { testRenderer, container, context };
  });
}

it("renders webhooks", async () => {
  const { container } = await createTestRenderer();
  await act(async () => {
    await wait(() => {
      expect(within(container).toJSON()).toMatchSnapshot();
    });
  });
});

it("goes to add new webhook endpoint when clicking add", async () => {
  const {
    container,
    context: { transitionControl },
  } = await createTestRenderer();

  // Prevent router transitions.
  transitionControl.allowTransition = false;

  act(() => {
    within(container)
      .getByTestID("add-webhook-endpoint")
      .props.onClick({ button: 0, preventDefault: noop });
  });

  // Expect a routing request was made to the right url.
  await act(async () => {
    await wait(() => {
      expect(transitionControl.history[0].pathname).toBe(
        "/admin/configure/webhooks/add"
      );
    });
  });
});
