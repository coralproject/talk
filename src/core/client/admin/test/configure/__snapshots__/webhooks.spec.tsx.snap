// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a list of webhook endpoints that have been configured 1`] = `
<div
  className="Box-root HorizontalGutter-root HorizontalGutter-double"
  data-testid="webhooks-container"
>
  <div
    className="Box-root ConfigBox-root"
  >
    <div
      className="Box-root Flex-root ConfigBox-title Flex-flex Flex-justifySpaceBetween"
    >
      <div>
        <label
          className="Header-root"
          htmlFor="configure-webhooks-header.title"
        >
          Configure webhook endpoint
        </label>
      </div>
      <div />
    </div>
    <div
      className="ConfigBox-content"
    >
      <div
        className="Box-root HorizontalGutter-root HorizontalGutter-spacing-4"
      >
        <p
          className="FormFieldDescription-root"
        >
          Configure an endpoint to send events to when events occur within
Coral. These events will be JSON encoded and signed. To learn more
about webhook signing, visit our 
          <a
            className="ExternalLink-root"
            href="https://docs.coralproject.net/coral/v5/integrating/webhooks/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Webhook Guide
          </a>
          .
        </p>
        <a
          className="Button-root Button-sizeRegular Button-colorRegular Button-variantRegular Button-uppercase Button-iconLeft Button-linkButton"
          data-color="regular"
          data-variant="regular"
          href="/admin/configure/webhooks/add"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="Icon-root Icon-md"
          >
            add
          </i>
          Add webhook endpoint
        </a>
        <h3
          className="Subheader-root"
        >
          Endpoints
        </h3>
        <table
          className="Table-root Table-fullWidth"
        >
          <thead
            className="TableHead-root"
          >
            <tr
              className="TableRow-root"
            >
              <th
                className="TableCell-root TableCell-header"
              >
                URL
              </th>
              <th
                className="TableCell-root TableCell-header"
              >
                Status
              </th>
              <th
                className="TableCell-root TableCell-header"
              />
            </tr>
          </thead>
          <tbody
            className="TableBody-root"
          >
            <tr
              className="TableRow-root TableRow-body"
              data-testid="webhook-endpoint-webhook-endpoint-1"
            >
              <td
                className="TableCell-root WebhookEndpointRow-urlColumn TableCell-body"
              >
                http://example.com/webhook-endpoint-1
              </td>
              <td
                className="TableCell-root TableCell-body"
              >
                <span
                  className="Marker-root StatusMarker-success Marker-colorPending Marker-variantRegular"
                >
                  Enabled
                </span>
              </td>
              <td
                className="TableCell-root TableCell-body"
              >
                <div
                  className="Box-root Flex-root Flex-flex Flex-justifyFlexEnd"
                >
                  <a
                    className="Button-root Button-sizeRegular Button-colorRegular Button-variantText Button-uppercase Button-iconRight Button-linkButton"
                    data-color="regular"
                    data-variant="text"
                    href="/admin/configure/webhooks/endpoint/webhook-endpoint-1"
                    onClick={[Function]}
                  >
                    Details 
                    <i
                      aria-hidden="true"
                      className="Icon-root Icon-sm"
                    >
                      keyboard_arrow_right
                    </i>
                  </a>
                </div>
              </td>
            </tr>
            <tr
              className="TableRow-root TableRow-body"
              data-testid="webhook-endpoint-webhook-endpoint-2"
            >
              <td
                className="TableCell-root WebhookEndpointRow-urlColumn TableCell-body"
              >
                http://example.com/webhook-endpoint-2
              </td>
              <td
                className="TableCell-root TableCell-body"
              >
                <span
                  className="Marker-root StatusMarker-error Marker-colorPending Marker-variantRegular"
                >
                  Disabled
                </span>
              </td>
              <td
                className="TableCell-root TableCell-body"
              >
                <div
                  className="Box-root Flex-root Flex-flex Flex-justifyFlexEnd"
                >
                  <a
                    className="Button-root Button-sizeRegular Button-colorRegular Button-variantText Button-uppercase Button-iconRight Button-linkButton"
                    data-color="regular"
                    data-variant="text"
                    href="/admin/configure/webhooks/endpoint/webhook-endpoint-2"
                    onClick={[Function]}
                  >
                    Details 
                    <i
                      aria-hidden="true"
                      className="Icon-root Icon-sm"
                    >
                      keyboard_arrow_right
                    </i>
                  </a>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`goes to add new webhook endpoint when clicking add 1`] = `
<div
  className="Box-root HorizontalGutter-root HorizontalGutter-double"
  data-testid="webhooks-container"
>
  <div
    className="Box-root ConfigBox-root"
  >
    <div
      className="Box-root Flex-root ConfigBox-title Flex-flex Flex-justifySpaceBetween"
    >
      <div>
        <label
          className="Header-root"
          htmlFor="configure-webhooks-header.title"
        >
          Configure webhook endpoint
        </label>
      </div>
      <div />
    </div>
    <div
      className="ConfigBox-content"
    >
      <div
        className="Box-root HorizontalGutter-root HorizontalGutter-spacing-4"
      >
        <p
          className="FormFieldDescription-root"
        >
          Configure an endpoint to send events to when events occur within
Coral. These events will be JSON encoded and signed. To learn more
about webhook signing, visit our 
          <a
            className="ExternalLink-root"
            href="https://docs.coralproject.net/coral/v5/integrating/webhooks/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Webhook Guide
          </a>
          .
        </p>
        <a
          className="Button-root Button-sizeRegular Button-colorRegular Button-variantRegular Button-uppercase Button-iconLeft Button-linkButton"
          data-color="regular"
          data-variant="regular"
          href="/admin/configure/webhooks/add"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="Icon-root Icon-md"
          >
            add
          </i>
          Add webhook endpoint
        </a>
        <h3
          className="Subheader-root"
        >
          Endpoints
        </h3>
        <div
          className="CallOut-root CallOut-colorRegular CallOut-fullWidth"
        >
          <div>
            There are no webhook endpoints configured, add one above.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`goes to the webhook endpoint configuration page when selected 1`] = `
<div
  className="Box-root HorizontalGutter-root HorizontalGutter-double"
  data-testid="webhooks-container"
>
  <div
    className="Box-root ConfigBox-root"
  >
    <div
      className="Box-root Flex-root ConfigBox-title Flex-flex Flex-justifySpaceBetween"
    >
      <div>
        <label
          className="Header-root"
          htmlFor="configure-webhooks-header.title"
        >
          Configure webhook endpoint
        </label>
      </div>
      <div />
    </div>
    <div
      className="ConfigBox-content"
    >
      <div
        className="Box-root HorizontalGutter-root HorizontalGutter-spacing-4"
      >
        <p
          className="FormFieldDescription-root"
        >
          Configure an endpoint to send events to when events occur within
Coral. These events will be JSON encoded and signed. To learn more
about webhook signing, visit our 
          <a
            className="ExternalLink-root"
            href="https://docs.coralproject.net/coral/v5/integrating/webhooks/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Webhook Guide
          </a>
          .
        </p>
        <a
          className="Button-root Button-sizeRegular Button-colorRegular Button-variantRegular Button-uppercase Button-iconLeft Button-linkButton"
          data-color="regular"
          data-variant="regular"
          href="/admin/configure/webhooks/add"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="Icon-root Icon-md"
          >
            add
          </i>
          Add webhook endpoint
        </a>
        <h3
          className="Subheader-root"
        >
          Endpoints
        </h3>
        <table
          className="Table-root Table-fullWidth"
        >
          <thead
            className="TableHead-root"
          >
            <tr
              className="TableRow-root"
            >
              <th
                className="TableCell-root TableCell-header"
              >
                URL
              </th>
              <th
                className="TableCell-root TableCell-header"
              >
                Status
              </th>
              <th
                className="TableCell-root TableCell-header"
              />
            </tr>
          </thead>
          <tbody
            className="TableBody-root"
          >
            <tr
              className="TableRow-root TableRow-body"
              data-testid="webhook-endpoint-webhook-endpoint-1"
            >
              <td
                className="TableCell-root WebhookEndpointRow-urlColumn TableCell-body"
              >
                http://example.com/webhook-endpoint-1
              </td>
              <td
                className="TableCell-root TableCell-body"
              >
                <span
                  className="Marker-root StatusMarker-success Marker-colorPending Marker-variantRegular"
                >
                  Enabled
                </span>
              </td>
              <td
                className="TableCell-root TableCell-body"
              >
                <div
                  className="Box-root Flex-root Flex-flex Flex-justifyFlexEnd"
                >
                  <a
                    className="Button-root Button-sizeRegular Button-colorRegular Button-variantText Button-uppercase Button-iconRight Button-linkButton"
                    data-color="regular"
                    data-variant="text"
                    href="/admin/configure/webhooks/endpoint/webhook-endpoint-1"
                    onClick={[Function]}
                  >
                    Details 
                    <i
                      aria-hidden="true"
                      className="Icon-root Icon-sm"
                    >
                      keyboard_arrow_right
                    </i>
                  </a>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders webhooks 1`] = `
<div
  className="Box-root HorizontalGutter-root HorizontalGutter-double"
  data-testid="webhooks-container"
>
  <div
    className="Box-root ConfigBox-root"
  >
    <div
      className="Box-root Flex-root ConfigBox-title Flex-flex Flex-justifySpaceBetween"
    >
      <div>
        <label
          className="Header-root"
          htmlFor="configure-webhooks-header.title"
        >
          Configure webhook endpoint
        </label>
      </div>
      <div />
    </div>
    <div
      className="ConfigBox-content"
    >
      <div
        className="Box-root HorizontalGutter-root HorizontalGutter-spacing-4"
      >
        <p
          className="FormFieldDescription-root"
        >
          Configure an endpoint to send events to when events occur within
Coral. These events will be JSON encoded and signed. To learn more
about webhook signing, visit our 
          <a
            className="ExternalLink-root"
            href="https://docs.coralproject.net/coral/v5/integrating/webhooks/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Webhook Guide
          </a>
          .
        </p>
        <a
          className="Button-root Button-sizeRegular Button-colorRegular Button-variantRegular Button-uppercase Button-iconLeft Button-linkButton"
          data-color="regular"
          data-variant="regular"
          href="/admin/configure/webhooks/add"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="Icon-root Icon-md"
          >
            add
          </i>
          Add webhook endpoint
        </a>
        <h3
          className="Subheader-root"
        >
          Endpoints
        </h3>
        <div
          className="CallOut-root CallOut-colorRegular CallOut-fullWidth"
        >
          <div>
            There are no webhook endpoints configured, add one above.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
